Readme.md
<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
      <ul>
        <li><a href="#built-with">Built With</a></li>
      </ul>
    </li>
    <li>
      <a href="#getting-started">Getting Started</a>
      <ul>
        <li><a href="#prerequisites">Prerequisites</a></li>
        <li><a href="#installation">Installation</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
  </ol>
</details>


<!-- ABOUT THE PROJECT -->
## About The Project
This project is for the llama2 chatbot built specifically for Mount Holyoke College information. 

### Built With
* ![Python](https://img.shields.io/badge/Python-3776AB?style=for-the-badge&logo=python&logoColor=white)
* ![JupyterLab](https://img.shields.io/badge/JupyterLab-F37626.svg?&style=for-the-badge&logo=Jupyter&logoColor=white)
* ![Llama.cpp](https://img.shields.io/badge/Llama.cpp-002b36?style=for-the-badge)
* ![Streamlit](https://img.shields.io/badge/Streamlit-FF4B4B?style=for-the-badge&logo=Streamlit&logoColor=white)
* ![Docker](https://img.shields.io/badge/Docker-2496ED?style=for-the-badge&logo=Docker&logoColor=white)
* ![Epsilla](https://img.shields.io/badge/Epsilla-007ACC?style=for-the-badge)
* ![BeautifulSoup](https://img.shields.io/badge/BeautifulSoup-8A2BE2?style=for-the-badge)
  
<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- GETTING STARTED -->
## Prerequisite

### Install Architecture
If you are using Apple Silicon (M1) Mac, make sure you have installed a version of Python that supports arm64 architecture; Otherwise, while installing it will build the llama.ccp x86 version which will be 10x slower on Apple Silicon (M1) Mac. To install arm64 architecture on your laptop, run the following code on your laptop terminal:
```zsh
arm64path = "Miniforge3-MacOSX-arm64.sh"
wget https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-MacOSX-arm64.sh
bash Miniforge3-MacOSX-arm64.sh
```

## Installation
### Clone the repo
```zsh
git clone https://github.com/AstridZhao/moho_bot.git
```
### Install essential packages
To have all the packages needed to run the code, you can run the below code in the **terminal** with the main program directory:
```zsh
pip install -r requirements.txt
```
<p align="right">(<a href="#readme-top">back to top</a>)</p>

### Install Llama.cpp
The main goal of llama.cpp is to run the LLaMA model using 4-bit integer quantization on a MacBook. The instructions below are for Macs with an **M1 chip**.
For other operating systems, you can find instructions [here](https://github.com/TrelisResearch/llamacpp-install-basics/blob/main/instructions.md).

Run the below code in **terminal**. Make sure the current directory should be the main program directory.
```zsh
git clone https://github.com/ggerganov/llama.cpp
cd llama.cpp
LLAMA_METAL=1 make
cd ..
```
Then, after installing llama.cpp, we can require the specific llama model from [huggingface](https://huggingface.co/TheBloke/Llama-2-7B-Chat-GGUF). In this project, we chose to use  "llama-2-7b-chat.Q4_K_M.gguf". Run the below code in the **terminal**.
```zsh
cd llama.cpp
wget https://huggingface.co/TheBloke/Llama-2-7b-Chat-GGUF/resolve/main/llama-2-7b-chat.Q4_K_M.gguf
cd ..
```

Now you have the llama model available to run on your laptop. 
To test if your installation is successful, you can do it with the below code in the **terminal**:
```zsh
cd llama.cpp
./main -m llama-2-7b-chat.Q4_K_M.gguf -c 1024 -ngl 8 -p "Where is new york?"
```

### Install Docker Desktop

In this project, we need to use the vector database (epsilla/vectordb), which could be pulled by docker.

Docker Desktop is a one-click-install application for your Mac, Linux, or Windows environment that lets you build, share, and run containerized applications and microservices. Docker packages software into standardized units called containers that have everything the software needs to run including libraries, system tools, code, and runtime.

You find the more detailed install instructions of docker [here](https://docs.docker.com/desktop/install/mac-install/).
I suggest you install the docker in the "/Users/{your username}" directory.

After installing successfully, you need to run the below code in **terminal** before initiating the program code each time.

```zsh
cd /Users/{your username}
open -a Docker
docker pull epsilla/vectordb
docker run --pull=always -d -p 8888:8888 epsilla/vectordb
```
<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- USAGE -->
## Program Initiation

To run this program project, you need to do FOUR steps: 

* 1. Initiate Docker with the command
     
     ```zsh
      cd /Users/{your username}
      open -a Docker
      docker pull epsilla/vectordb
      docker run --pull=always -d -p 8888:8888 epsilla/vectordb
    ```
and then return back to the program directory.
  
* 2. To extract text from the Mount Holyoke College website, run [collect.py].
     You can find or edit the website link in "URL.txt" file in the folder.

      ```zsh
     python3 collect.py
     ```
     
* 3. To train the dataset, run [learn.py] by using command
     
     ```zsh
     python3 learn.py
     ```
  
* 4. To run the chatbot, run [app.py] by using command

     ```zsh
     streamlit run app.py
     ```

**I suggest you run the program following the above sequence to avoid unexpected errors.**

<p align="right">(<a href="#readme-top">back to top</a>)</p>

## Chatbot Interface

After running the app.py, the streamlit will bring you to the chatbot interface.


Then, you can ask questions to MohoBot. 

https://github.com/AstridZhao/moho_bot/assets/79214456/5bde6326-8039-458f-9949-a4c0ad573064




Documents_collection/history.txt
Skip to main site navigationSkip to main contentDirectoryNews & StoriesAthleticsEventsInformation ForCurrent StudentsAlumsParents & FamiliesFacultyStaffFive College StudentsNew StudentsGiveAcademicsFind Your ProgramSTEM at Mount HolyokeThe Arts at Mount HolyokeEnhance Your LearningInternational LearningAfter Mount HolyokeAcademic Centers and FacilitiesAcademic Support and ResourcesFaculty MentorshipResearch Excellence and OpportunityDegrees and Graduation RequirementsFaculty AccomplishmentsAdmission & Financial AidUndergraduateGraduateVisiting or Connecting with Mount HolyokeInclusive AdmissionStudent ExperienceLiving on CampusCampus ExperienceCommunity and BelongingBeyond CampusAthleticsWhy Mount HolyokeAlum Network and SupportAt a GlanceDiversity, Equity and InclusionMission and AdministrationNews & StoriesSustainabilityWe Are Mount HolyokeWhy a Women's College?SearchAcademicsFind Your ProgramHistoryHistoryUndergraduateHistory is a critical and analytical method of inquiry into our collective past based on our cumulative experience, informed understanding, and careful judgment.History FacultyHistory ResearchProgram OverviewUsing evidence from the past, historians investigate the relationship between cause and effect and between agency and responsibility to understand better how things change over time.These general habits of sorting through the past allow us to appreciate the profound differences between ourselves and others and to imagine (and to some degree experience) the world as people have in times now lost and in places we shall never see.If you wish to major in history you would typically declare your intention sometime in the fourth semester of College.The requirements for a major in history emphasize both a broad exposure to different time periods and regions of the world, as well as a rigorous engagement with the practice of history as a discipline grounded in the use of primary sources.Historians can be found in many guises, including in other departments:Geoffrey S. Sumi, Professor of Classics,Lauret Savoy, Professor of Environmental Studies andPaula Debnar, Professor of Classics are all faculty affiliated with the history department.Takinghistory courses offered through the five college campusesand theHampshire Colleges History Department,Amherst College Department of History,Smith College Department of Historyand theUniversity of Massachusetts Amherst Department of Historyalso offer many additional opportunities.Program InformationProgram TypeMajorMinorOutcomeBachelor of ArtsDepartmentDepartment of HistoryWhy study history?There are as many answers to this question as there are stories to be told of the past, but the American Historical Association’s“Tuning Project”– developed with the help of Mount Holyoke College history alum professor Anne Hyde ‘82 — offers a clear and compelling statement on the value and skills of historical thinking.Community VoicesSpotlight on History majorsGo to the previous slide.Go to the next slide.Gaining confidence to pursue opportunities like learning a new language and doing an independent studyAnsley Keane ’23she/herFollowing passions leads to a career in helping others address traumaLaura Khoudari ’00“A symbiosis of having fun and getting the work done”Rebeca Villatoro-Alvarez ’23Explore more storiesOur coursesWe offer courses at several levels.More advanced levels do often involve heavier reading and writing assignments, but critical reading and clear writing are valued in all history courses, and the distinction between levels is often as much about the kinds of questions asked as the level of rigor involved.Many courses can be taken without any prerequisites.First Year Seminarsare only open to first-year students and introduce history as a field of inquiry and stress the development of basic writing, arguing and research methods.100-level regional surveysfocus on particular geographic areas; they have no prerequisites and are open to all students.200-level coursesexamine particular times and places in history and range in size from large survey courses to small seminars.In most cases, all students are welcome to enroll.300-level courses(colloquium and research seminars) offer opportunities to examine topics in depth; some do have prerequisites or require permission of the instructor.If permission is required, please complete and submit the 300-Level Application Form.Take one history course and you will undoubtedly want to take more!Selecting your first history courseYou may prefer to start with the kind of broad introduction available in a survey, to learn about the diversity of peoples, places, and ways of living that have come before us and out of which the world as we know it has emerged.Or you may prefer to dive into an area of history that seems particularly intriguing or relevant to your interests, and perhaps turn to one or another survey later to gain a broader understanding of the past.Since history always involves the interplay between broad contexts and particular places in time, you can begin on either end of the spectrum.You are not required to begin at the 100 level.Selecting courses in your first yearThese fall courses are recommended for first-year students:SurveysHIST-124-01 History of Modern South Asia, 1700 to the PresentHIST-137-01 Modern East Asia, 1600-2000HIST-141-01 Introduction to Modern African HistoryHIST-142-01 Introduction to Pre-colonial African HistoryHIST-151-01 Modern & Contemporary EuropeHIST-171-01 The American Peoples Since 1865HIST-180-01 Introduction to Latin American CulturesCourses and RequirementsGenerally speaking, history courses are reading and writing intensive.To succeed as a student of history, you must, therefore learn to read actively and write clearly.Learning GoalsMajor RequirementsMinor RequirementsCoursesLearning GoalsLearning GoalsStudents of the History department can expect to:Understand the dynamics of change over time, the complexity of human experience across time and space, a

learn.py
import os
import ssl
# from nltk.tokenize import word_tokenize, sent_tokenize
# from nltk.stem.snowball import PorterStemmer
# from nltk.corpus import stopwords
from langchain.document_loaders import TextLoader 
from langchain.text_splitter import CharacterTextSplitter, RecursiveCharacterTextSplitter
from langchain.vectorstores import Epsilla
from sentence_transformers import SentenceTransformer
from typing import List
from glob import glob
from langchain.vectorstores import Epsilla
from pyepsilla import vectordb
from langchain.vectorstores import Chroma

# try:
#     _create_unverified_https_context = ssl._create_unverified_context
# except AttributeError:
#     pass
# else:
#     ssl._create_default_https_context = _create_unverified_https_context
# nltk.download('punkt')
# nltk.download('averaged_perceptron_tagger')


#run in terminal:
# if (os.getcwd() != "/Users/astridz"):
#     os.chdir('/Users/astridz')
# docker pull epsilla/vectordb
# docker run --pull=always -d -p 8888:8888 epsilla/vectordb

def reading(fileName):
    f = open(fileName, "r")
    content = f.read() 
    f.close()
    return content

# Local embedding model
model = SentenceTransformer('all-MiniLM-L6-v2')

class LocalEmbeddings():
    def embed_documents(self, texts: List[str]) -> List[List[float]]:
        return model.encode(texts).tolist()

embeddings = LocalEmbeddings()
files = glob("./Documents_collection/*")

# os.chdir("./Documents_collection/")

# files = ["statistics.txt"]

splitted_documents = []
splitter = RecursiveCharacterTextSplitter( separators=[" ", ",", "\n"],chunk_size=1000, chunk_overlap=200)

for file in files:
    loader = TextLoader(file)
    documents = loader.load()
    split_docs = splitter.split_documents(documents)
    print("Splitted document chunk size for current file:", len(split_docs))
    splitted_documents.extend(split_docs)
    # print("splited document:" , len(splitted_documents))

client = vectordb.Client()

try:
    vector_store = Epsilla.from_documents(
        splitted_documents,
        embeddings,
        client,
        db_path="/tmp/localchatdb",
        db_name="LocalChatDB",
        collection_name="LocalChatCollection"
    )
except Exception as e:
    print("An error occurred:", e)



collect.py
import sys
import os
import nltk
import ssl
from nltk.tokenize import word_tokenize, sent_tokenize
from transformers import AutoTokenizer, AutoModel
from operator import itemgetter
import string
from nltk.stem.snowball import PorterStemmer
from nltk.corpus import stopwords
from collections import Counter
from scipy.interpolate import make_interp_spline
import numpy as np
import pandas as pd
from langchain.document_loaders import TextLoader
from langchain.text_splitter import CharacterTextSplitter, RecursiveCharacterTextSplitter
from langchain.vectorstores import Epsilla
from pyepsilla import vectordb
from sentence_transformers import SentenceTransformer
from typing import List
from glob import glob

import ssl
try:
    _create_unverified_https_context = ssl._create_unverified_context
except AttributeError:
    pass
else:
    ssl._create_default_https_context = _create_unverified_https_context
# nltk.download('punkt')
# nltk.download('averaged_perceptron_tagger')


def reading(fileName):
    f = open(fileName, "r")
    content = f.read() 
    f.close()
    return content

def writing(input, filename):
    f = open(filename, 'w')
    f.write(input)
    f.close()
    
# # Method1: ------------------------------------->
# from langchain.document_loaders import WebBaseLoader
# webs = ['https://www.mtholyoke.edu/academics/find-your-program/computer-science']


# datas = []
# for web in webs:
#     loader = WebBaseLoader(web)
#     data = loader.load()
#     datas.append(data[0].page_content)

# if (os.getcwd() != "/Users/astridz/Documents/Moho_Bot/Documents_collection"):
#     os.chdir('Documents_collection')
    
# for n, data in enumerate(datas):
#     writing(data, f"page{n}.txt")
    
# os.chdir('../')

# # ------------------------------------->

from bs4 import BeautifulSoup
from bs4.element import Comment
import urllib.request

def tag_visible(element):
    if element.parent.name in ['style', 'script', 'head', 'title', 'meta', '[document]']:
        return False
    if isinstance(element, Comment):
        return False
    return True

def text_from_html(body):
    soup = BeautifulSoup(body, 'html.parser')
    texts = soup.findAll(string=True)
    visible_texts = filter(tag_visible, texts)  
    return u"".join(t.strip() for t in visible_texts)

URLs = reading("URL.txt")
URLList = URLs.split()

if (os.getcwd() != "/Users/astridz/Documents/Moho_Bot/Documents_collection"):
        os.chdir('Documents_collection')
for link in URLList:
    pure_name = link.split('/')[-1]
    html = urllib.request.urlopen(link).read()
    document = text_from_html(html)
    document_tokenize = "".join(nltk.sent_tokenize(document))
    writing(document_tokenize, f"{pure_name}.txt")





testing_command.py
import shlex, subprocess, os
os.chdir('/Users/astridz/Documents/llama.cpp')
command_line = input()
# ./main -m llama-2-7b-chat.Q4_K_M.gguf -c 2048 -c N 2048 -ngl 48 -p
# ./main -m llama-2-7b-chat.Q4_K_M.gguf -c 2048 -ngl 48 -p
# command = ['./main', '-m', pure_name, '-c', '2048', '-ngl', '48', '-n N', '5',  '-e', '-p', prompt]
# ./main -m llama-2-7b-chat.Q4_K_M.gguf -c 2048 -ngl 48 -p "You are a helpful question answer assistant. Based on given Context: comsc150 followed comsc161 combination two equivalent comsc151for student curious computer science comsc151ds comsc151hc 2 comsc150 plus comsc121 substitute comsc151 3 comsc205py plus comsc122 COMSC-161 serves as an alternate prerequisite route for COMSC-205 Data Structures.COMSC-205  Data computer science study, COMSC-150 should be followed by COMSC-161; the combination of the two is comsc161 serves alternate prerequisite route comsc205 data structurescomsc205 data structuresfall, answer question what is COMSC151?""

"""
        You are a helpful question answer assistant. Based on given Context: comsc150 followed comsc161 combination two equivalent comsc151for student curious computer science comsc151ds comsc151hc 2 comsc150 plus comsc121 substitute comsc151 3 comsc205py plus comsc122 COMSC-161 serves as an alternate prerequisite route for COMSC-205 Data Structures.COMSC-205  Data computer science study, COMSC-150 should be followed by COMSC-161; the combination of the two is comsc161 serves alternate prerequisite route comsc205 data structurescomsc205 data structuresfall, answer question what is COMSC151?. 
        
        Do not copy the context in your answer.
        Try to understand the Context and rephrase them.
        Please don't make things up or say things not mentioned in the Context. 
        
"""
args = shlex.split(command_line)
print(args)

p = subprocess.Popen(args,stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
output = p.stdout.readlines()
# try:
#     # print("communicated")
#     outs, errs = p.communicate(timeout=15)
#     print("communicated")
# except subprocess.TimeoutExpired:
#     p.kill()
#     outs, errs = p.communicate()

print("Answer: \n " )
print(output)

os.chdir('/Users/astridz/Documents/Moho_Bot')

testing_cttransformer.py
from langchain.prompts import PromptTemplate
from langchain.chains import LLMChain
from langchain.llms import CTransformers
import os

model_name = 'TheBloke/Llama-2-7b-Chat-GGUF/resolve/main/llama-2-7b-chat.Q4_K_M.gguf'
pure_name = model_name.split('/')[-1]

llamacpp_directory = '/Users/astridz/Documents/llama.cpp'
local_directory = '/Users/astridz/Documents/moho_bot'
docker_directory = '/Users/astridz'


DEFAULT_PROMPT = "You are a helpful, respectful and honest assistant."
instruction = "Convert the following text from English to Chinese: {text}."
text = "you are beautiful."
B_INST, E_INST = "[INST]", "[/INST]"
B_SYS, E_SYS = "<<SYS>>\n", "\n<</SYS>>\n\n"
DEFAULT_PROMPT = B_SYS + DEFAULT_PROMPT  + E_SYS
template = B_INST + DEFAULT_PROMPT + instruction + E_INST
# template = B_INST + DEFAULT_PROMPT + instruction + E_INST

prompt = PromptTemplate(template=template,input_variables= ["text"])

os.chdir(llamacpp_directory)
llm = CTransformers(model=pure_name,
                model_type='llama',
                config={'max_new_tokens': 128,
                        'temperature': 0.01}
                )

LLM_Chain=LLMChain(prompt=prompt, llm=llm)
print(LLM_Chain.run("How are you"))

Past_trials/testing_dockercommand.py
import subprocess, os
from langchain.vectorstores import Epsilla

if (os.getcwd() != "/Users/astridz"):
    os.chdir('/Users/astridz')
# Pull the Docker image

start_command = "open -a Docker"
start_process = subprocess.Popen(start_command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
out, err = start_process.communicate()
if start_process.returncode != 0:
    print(f"Error start Docker: {err}")
else:
    print("start success! \n")
    print(f"Output: {out} \n")

pull_command = "docker pull epsilla/vectordb"
pull_process = subprocess.Popen(pull_command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
out, err = pull_process.communicate()
if pull_process.returncode != 0:
    print(f"Error pulling Docker image: {err}")
else:
    print("pull success! \n")
    print(f"Output: {out} \n")

# Run the Docker container
run_command = "docker run --pull=always -d -p 8888:8888 epsilla/vectordb"
run_process = subprocess.Popen(run_command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
out, err = run_process.communicate()
if run_process.returncode != 0:
    print(f"Error running Docker container: {err}")
else:
    print("r! success! \n")
    print(f"Output: {out} \n")
    
os.chdir('/Users/astridz/Documents/moho_bot')

app.py

from pyepsilla import vectordb
from sentence_transformers import SentenceTransformer
import streamlit as st
import os
import subprocess, threading
from typing import List
from transformers import AutoTokenizer
import requests
import json
import time, queue

import subprocess, os
from langchain.vectorstores import Epsilla
from langchain.prompts import PromptTemplate
# start docker ->---------------->---------------->---------------
# if (os.getcwd() != "/Users/astridz"):
#     os.chdir('/Users/astridz')
# # Pull the Docker image

# start_command = "open -a Docker"
# subprocess.run(start_command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
# # out, err = start_process.communicate()
# # if start_process.returncode != 0:
# #     print(f"Error start Docker: {err}")
# # else:
# #     print("start success! \n")
# #     print(f"Output: {out} \n")

# pull_command = "docker pull epsilla/vectordb"
# subprocess.run(pull_command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
# # out, err = pull_process.communicate()
# # if pull_process.returncode != 0:
# #     print(f"Error pulling Docker image: {err}")
# # else:
# #     print("pull success! \n")
# #     print(f"Output: {out} \n")

# # # Run the Docker container
# run_command = "docker run --pull=always -d -p 8888:8888 epsilla/vectordb"
# subprocess.run(run_command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
# # out, err = run_process.communicate()
# # if run_process.returncode != 0:
# #     print(f"Error running Docker container: {err}")
# # else:
# #     print("r! success! \n")
# #     print(f"Output: {out} \n")
    
# os.chdir('/Users/astridz/Documents/moho_bot')

# ->---------------->---------------->---------------

local_directory = "/Users/astridz/Documents/moho_bot"
# Local embedding model for embedding the question.
model = SentenceTransformer('all-MiniLM-L6-v2')
class LocalEmbeddings():
    def embed_query(self, text: str) -> List[float]:
        return model.encode(text).tolist()
    
embeddings = LocalEmbeddings()
# Connect to Epsilla as knowledge base.
client = vectordb.Client()
vector_store = Epsilla(
    client,
    embeddings,
    db_path="/tmp/localchatdb",
    db_name="LocalChatDB"
)

vector_store.use_collection("LocalChatCollection")

model_name = 'TheBloke/Llama-2-7b-Chat-GGUF/resolve/main/llama-2-7b-chat.Q4_K_M.gguf'
pure_name = model_name.split('/')[-1]

st.set_page_config(page_title="🦙💬 Llama 2 Chatbot with Streamlit")

#Create a Side bar
with st.sidebar:
    st.title("🦙💬 Llama 2 MohoBot")
    st.header("Settings")
    
# Clear the Chat Messages
def clear_chat_history():
    st.session_state.messages=[{"role":"assistant", "content": "As a helpful question answer assistant, how can I help you today?"}]
    print_dialog()
# create Clear button
st.sidebar.button('Clear Chat History', on_click=clear_chat_history)

# Diplay the chat messages
def print_dialog(): 
    for message in st.session_state.messages:
        with st.chat_message(message["role"]):
            st.write(message["content"])
    
    
#store message 
if "messages" not in st.session_state.keys():
    st.session_state.messages=[{"role": "assistant", "content": "As a helpful question answer assistant, how can I help you today?"}]
    # st.session_state["messages"] = [{"role": "assistant", "content": "How can I help you?"}]
    st.chat_message("assistant").write("As a helpful question answer assistant, how can I help you today?")


if question := st.chat_input():
    question = question.strip()
    st.session_state.messages.append({"role": "user", "content": question})
    print_dialog()
    #vector_store.similarity_search
    context = ''.join(map(lambda doc: doc.page_content, vector_store.similarity_search(question, k = 2)))
    # Split into lines, strip whitespace, filter out empty lines, and join back into a single string
    clean_context = '\n'.join([line.strip() for line in context.splitlines() if line.strip()])
    
    SYSTEM_PROMPT =f'''As a helpful question answer assistant, answer the user question by giving response and follow the rules: \
    1. Do not copy the context in your answer. Try to understand the context and rephrase them. \
    2. Please don't make things up or say things not mentioned in the Context. \
    3. You can trust the context. \
    The context is: {clean_context}'''
    # prompt_template = f'''[INST]<<SYS>>\n{SYSTEM_PROMPT}{context}\n<</SYS>>\n\n{question}[/INST]'''
    # prompt_template = f'''[INST]<<SYS>>\n{SYSTEM_PROMPT}{context}\n<</SYS>>\n\n{question}[/INST]'''
    B_INST, E_INST = "[INST]", "[/INST]"
    B_SYS, E_SYS = "<<SYS>>\n", "\n<</SYS>>\n\n"
    SYSTEM_PROMPT = B_SYS + SYSTEM_PROMPT + E_SYS
    prompt_template = B_INST + SYSTEM_PROMPT + question + E_INST

    # Start the subprocess and the threading to handle its output
    if (os.getcwd() != '/Users/astridz/Documents/llama.cpp'):
        os.chdir('/Users/astridz/Documents/llama.cpp')
    
    # prompt_template = f'''./main -m llama-2-7b-chat.Q4_K_M.gguf -c 1024 --verbose-prompt --keep 0 -b 16 -ngl 48 --prompt '{SYSTEM_PROMPT}{context} USER: {question}' --in-suffix 'ASSISTANT:' '''
    
    # ./main -m llama-2-7b-chat.Q4_K_M.gguf -c 1024 --verbose-prompt --keep 0 -b 16 -ngl 48 -p
    pure_name = "llama-2-7b-chat.Q4_K_M.gguf"
    print(prompt_template)
    command = ['./main', '-m', pure_name, '-c', '1024', '--multiline-input', '-ngl', '8', '-p', prompt_template]
    process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE,text=True)
    start_time = time.time()
    while True:
        if process.poll() is not None:
            break  # Break the loop if the subprocess has finished
        
        output = process.stdout.readline()
        assistant_response = ""
        if output:
            if '[/INST]' in output:
                inst_index = output.find('[/INST]')
                # Check if [/INST] is found in the text
                if inst_index != -1:
                    # Print everything after [/INST]
                    assistant_response = output[inst_index + len('[/INST]'):].strip()
                    st.session_state.messages.append({"role" : "assistant", "content": f'''{assistant_response}</s><s>[INST]'''})
                    st.chat_message("assistant").write(assistant_response)
            elif '<</SYS>>' in output or '[INST]' in output:
                continue
            elif  'As a helpful question answer assistant, ' in output:
                st.chat_message("assistant").write("processing...")
            else:
                assistant_response = output.strip()
                if assistant_response == "":
                    continue
                else:
                    st.session_state.messages.append({"role":"assistant", "content": f'''{assistant_response}</s><s>[INST]'''})
                    st.chat_message("assistant").write(assistant_response)
    
        # end_time = time.time()
        # elapsed_time = end_time - start_time
        # print("total time :",  elapsed_time)
        
        # process.stdout.close()

